shmem/man/energymon-shmem-provider.1.in
.TH "@MAN_BINARY_PREFIX_UPPER@\-SHMEM\-PROVIDER" "1" "2022-03-12" "energymon" "EnergyMon Utilities"
.SH "NAME"
.LP
@MAN_BINARY_PREFIX@\-shmem\-provider \- provide EnergyMon readings over
shared memory
.SH "SYNPOSIS"
.LP
\fB@MAN_BINARY_PREFIX@\-shmem\-provider\fP
[\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
Provide EnergyMon readings over shared memory, e.g., for use by
\fBlibenergymon\-shmem\fP.
Uses the @MAN_IMPL@ EnergyMon implementation.
.LP
Both the provider and the consumer must agree on the \fIpath\fP and \fIid\fP
used to create the IPC shared memory key, as required by \fBftok(3)\fP.
.LP
If the consumer is \fBlibenergymon\-shmem\fP, the shared memory provider must
be running before the energymon context is initialized.
To specify \fIpath\fP and \fIid\fP values for \fBlibenergymon\-shmem\fP, set
the \fBENERGYMON_SHMEM_DIR\fP and \fBENERGYMON_SHMEM_ID\fP environment
variables, respectively.
.SH "OPTIONS"
.LP
.TP
\fB\-h\fP, \fB\-\-help\fP
Prints the help screen.
.TP
\fB\-d\fP, \fB\-\-dir\fP=\fIPATH\fP
The shared memory path (default = ".").
.TP
\fB\-i\fP, \fB\-\-id\fP=\fIID\fP
The shared memory identifier (default = 1).
\fIID\fP must be in range [1, 255].
.SH "EXAMPLES"
.TP
\fB@MAN_BINARY_PREFIX@\-shmem\-provider\fP
Run the shared memory provider with default options.
.TP
\fB@MAN_BINARY_PREFIX@\-shmem\-provider \-d "/tmp/shmem" -i 10\fP
Run the shared memory provider with shared memory path "/tmp/shmem" and
shared memory identifier 10.
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/energymon/energymon>
.SH "SEE ALSO"
.BR ftok (3)

RELEASES.md
# Release Notes

## [Unreleased]

### Added

* ibmpowernv: new implementation for IBM PowerNV systems (targeted for OLCF Summit)
* ipg: new implementation using the Intel Power Gadget 3.7 API (PowerGadgetLib.h)
* jetson: new implementation for NVIDIA Jetson systems
* energymon-file-provider: -F/--force option to force updates faster than the EnergyMon implementation claims
* energymon-info: -h/--help option to print help message
* energymon-overhead: -h/--help option to print help message
* energymon-power-poller: -F/--force option to force updates faster than the EnergyMon implementation claims
* shmem: man pages for shared memory providers

### Fixed

* odroid, odroid-ioctl, zcu102: save errno when cleaning up after init failure
* osp: link with librt on older systems
* shmem: enforce provider ID in range `[1, 255]`
* wattsup-libftdi: compilation fails with clang when `-Werror` option is used
* utils: use implementation-specific binary names in help messages (don't assume energymon-default name format)


## [v0.4.0] - 2021-05-12

### Added

* ibmpowernv-power: new implementation for IBM PowerNV systems (targeted for OLCF Summit)
* raplcap-msr: support for multi-die packages
* Travis CI: test on additional architectures

### Changed

* CMake: overhaul the project build system
  * Increased minimum CMake version from 2.8.5 to 3.6
  * Refactor library and binary target creation
  * Export CMake targets during install
* rapl: document Linux kernel changes/differences in README
* raplcap-msr: update minimum dependency version to 0.5.0
* wattsup-libftdi [#47]: prefer `ftdi_tcioflush` if available (`ftdi_usb_purge_buffers` deprecated in libftdi 1.5)
* energymon-idle-power: now accepts idle time in fractional seconds

### Fixed

* Valid 0 return values from fread functions mistaken for errors when errno is already set
* cray-pm: building as default libraries used wrong macro, causing build failures
* osp: some cumulative energy unaccounted for during overflow detection (precision loss)
* raplcap-msr: Requires.private field in pkg-config file should specify the minimum version for raplcap-msr dependency
* wattsup-libftdi: Description field in pkg-config file was empty
* energymon-power-poller: incorrect Pmin summary output when number of samples is 0


## [v0.3.1] - 2020-10-27

### Fixed

* rapl [#46]: Check that top-level zones are actually package (not psys) zones
* rapl: Set errno when when no zones are found during init


## [v0.3.0] - 2020-03-10

### Added

* raplcap-msr: New implementation using libraplcap-msr (more features than `msr` and `rapl` implementations)
* utils [#44]: Build utilities for all implementations, not just energymon-default
* test: Build tests for all implementations, not just energymon-default

### Changed

* man pages: More pedantic formatting
* RELEASES: Reformat file

### Fixed

* cmake [#14]: Print useful CMake error when bad ENERGYMON_BUILD_DEFAULT value is set
* rapl: Powercap directory paths use lower case hex values, not decimal values
* zcu102: Potential memory leak in 'get_sensor_directories' under failure conditions


## [v0.2.4] - 2018-06-01

### Changed

* WattsUp implementations:
  * Wait up to 5 seconds for a good data packet (or fail) before completing initialization
  * Purge data buffers during initialization
  * Poll device at shorter intervals to increase energy value accuracy
  * Better handling of incomplete data packets
* osp-polling: Allow polling thread to be interrupted during sleep when finish is requested
* energymon-power-poller: Account for true elapsed time to produce more accurate power values

### Fixed

* wattsup-libftdi, wattsup-libusb: Fix empty reads by always setting baud rate and serial line properties
* wattsup-dev: Use pselect to avoid timer expired errors
* odroid, odroid-ioctl, osp-polling, zcu102: Minor correction to polling delay interval
* msr: Possible (but unlikely) precision computation bug
* osp, osp-polling: Fix status check during init that didn't always work


## [v0.2.3] - 2017-11-19

### Added

* Implementation for Xilinx ZCU102 (Jason Miller)

### Fixed

* Fixed [#40]: Time conversion bug in common function 'energymon_gettime_elapsed_us' (Jason Miller)


## [v0.2.2] - 2017-10-26

### Added

* New capabilities to some utility applications:
  * energymon-cmd-profile: Accept commands without quotations; add help option
  * energymon-file-provider: Add help, count, interval, and no-rewind options
  * energymon-idle-power: Add help option
* Man pages for utility applications
* VERSION and SOVERSION properties on shared object libraries
* Multiarch support (use GNU standard installation directories)

### Changed

* Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
* Refactor this RELEASES.md file

### Removed

* Removed private symbol exports in shared object libraries (prevents symbol name clashes)

### Fixed

* Various bug fixes to utility applications


## [v0.2.1] - 2017-10-01

### Added

* Support for msr-safe kernel module in msr implementation
* New cray-pm implementations to support capabilities of Cray XC30 and XC40 systems
* New wattsup-libusb and wattsup-libftdi implementations
* AppVeyor test configuration for Windows continuous integration

### Changed

* Faster refresh in osp-polling, odroid and odroid-ioctl implementations
* Better multi-platform support for some implementations, especially with portable time and sleep abstractions
* Downgraded required CMake version from 2.8.9 to 2.8
* Documented FSE 2016 and Technical Report publication information
* Various enhancements, optimizations, and documentation improvements

### Deprecated

* Deprecated custom DEFAULT CMake option in favor of ENERGYMON_BUILD_DEFAULT

### Removed

* Removed Android CMake toolchain file: cmake-toolchain/android.toolchain.cmake

### Fixed

* Fixed [#37]: In osp implementation, force overflow in ODROID Smart Power at 1000 Wh to avoid device counter saturation at 8192 Wh

shmem/energymon-shmem-provider.c
/**
 * Provide energymon readings over shared memory.
 *
 * @author Connor Imes
 * @date 2016-02-18
 */
#define _GNU_SOURCE
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <time.h>
#include "energymon.h"
#include "energymon-get.h"
#include "energymon-shmem.h"

#ifndef ENERGYMON_UTIL_PREFIX
#error Must set ENERGYMON_UTIL_PREFIX
#endif

static volatile int running = 1;
static energymon_shmem* ems;
static const char* key_dir = NULL;
static int key_proj_id = -1;
static int shm_id;

static const char short_options[] = "hd:i:";
static const struct option long_options[] = {
  {"help",      no_argument,       NULL, 'h'},
  {"dir",       required_argument, NULL, 'd'},
  {"id",        required_argument, NULL, 'i'},
  {0, 0, 0, 0}
};

static void print_usage(int exit_code) {
  fprintf(exit_code ? stderr : stdout,
          "Usage: "ENERGYMON_UTIL_PREFIX"-shmem-provider [OPTION]...\n\n"
          "Provide EnergyMon readings over shared memory, e.g., for use by\n"
          "libenergymon-shmem.\n\n"
          "Both the provider and the consumer must agree on the path and id used to create\n"
          "the IPC shared memory key, as required by ftok(3).\n\n"
          "If the consumer is libenergymon-shmem, the shared memory provider must be\n"
          "running before the energymon context is initialized. To specify path and id\n"
          "values for libenergymon-shmem, set the ENERGYMON_SHMEM_DIR and\n"
          "ENERGYMON_SHMEM_ID environment variables, respectively.\n\n"
          "Options:\n"
          "  -h, --help               Print this message and exit\n"
          "  -d, --dir=PATH           The shared memory path (default = \"%s\")\n"
          "  -i, --id=ID              The shared memory identifier (default = %d)\n"
          "                           ID must be in range [1, 255]\n",
          ENERGYMON_SHMEM_DIR_DEFAULT, ENERGYMON_SHMEM_ID_DEFAULT);
  exit(exit_code);
}

static void enforce_key_proj_id(void) {
  if (key_proj_id <= 0 || key_proj_id > 255) {
    print_usage(EINVAL);
  }
}

static void parse_args(int argc, char** argv) {
  const char* key_proj_id_env;
  int c;
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage(0);
        break;
      case 'd':
        key_dir = optarg;
        break;
      case 'i':
        key_proj_id = atoi(optarg);
        enforce_key_proj_id();
        break;
      case '?':
      default:
        print_usage(EINVAL);
        break;
    }
  }
  if (key_dir == NULL) {
    key_dir = getenv(ENERGYMON_SHMEM_DIR);
    if (key_dir == NULL) {
      key_dir = ENERGYMON_SHMEM_DIR_DEFAULT;
    }
  }
  if (key_proj_id < 0) {
    key_proj_id_env = getenv(ENERGYMON_SHMEM_ID);
    if (key_proj_id_env == NULL) {
      key_proj_id = ENERGYMON_SHMEM_ID_DEFAULT;
    } else {
      key_proj_id = atoi(key_proj_id_env);
      enforce_key_proj_id();
    }
  }
}

static void shandle(int sig) {
  switch (sig) {
    case SIGTERM:
    case SIGINT:
#ifdef SIGQUIT
    case SIGQUIT:
#endif
#ifdef SIGKILL
    case SIGKILL:
#endif
#ifdef SIGHUP
    case SIGHUP:
#endif
      running = 0;
    default:
      break;
  }
}

static int cleanup_shmem(void) {
  // detach shared memory
  if (shmdt(ems)) {
    perror("shmdt");
    return -errno;
  }
  // destroy shared memory
  if (shmctl(shm_id, IPC_RMID, NULL)) {
    perror("shmctl");
    return -errno;
  }
  return 0;
}

int main(int argc, char** argv) {
  energymon em;
  struct timespec ts;
  key_t mem_key;

  // register the signal handler
  signal(SIGINT, shandle);

  parse_args(argc, argv);

  // get the shared memory
  mem_key = ftok(key_dir, key_proj_id);
  shm_id = shmget(mem_key, sizeof(energymon_shmem), 0644 | IPC_CREAT | IPC_EXCL);
  if (shm_id < 0) {
    perror("shmget");
    return -errno;
  }
  ems = (energymon_shmem*) shmat(shm_id, NULL, 0);
  if (ems == (energymon_shmem*) -1) {
    perror("shmat");
    cleanup_shmem();
    return -errno;
  }

  // get the energy monitor
  if (energymon_get(&em)) {
    cleanup_shmem();
    return -errno;
  }

  // initialize the energy monitor
  if (em.finit(&em)) {
    perror("energymon:finit");
    cleanup_shmem();
    return -errno;
  }
  
  // store the interval in shared memory
  ems->interval_us = em.finterval(&em);
  ts.tv_sec = (time_t) (ems->interval_us / (uint64_t) 1000000);
  ts.tv_nsec = (long) ((ems->interval_us % (uint64_t) 1000000) * (uint64_t) 1000);

  // store the precision in shared memory
  ems->precision_uj = em.fprecision(&em);

  while (running) {
    // update the energy in shared memory
    ems->energy_uj = em.fread(&em);
    nanosleep(&ts, NULL);
  }

  errno = 0;
  // cleanup
  if (em.ffinish(&em)) {
    perror("energymon:ffinish");
  }

  cleanup_shmem();

  return errno;
}
